plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Mappings.hs

```diff
@@ -0,0 +1,174 @@
+                                                  SemanticTokens (SemanticTokens),
+                                                  UInt, absolutizeTokens)
+
+-- * 1. Mapping semantic token type to and from the LSP default token type.
+
+-- | map from haskell semantic token type to LSP default token type
+toLspTokenType :: HsSemanticTokenType -> SemanticTokenTypes
+toLspTokenType tk = case tk of
+  -- TVariable     -> SemanticTokenTypes_Variable
+  -- left hand side of none pattern bind
+  TFunction     -> SemanticTokenTypes_Function
+  TVariable     -> SemanticTokenTypes_Variable
+  TClass        -> SemanticTokenTypes_Class
+  TClassMethod  -> SemanticTokenTypes_Method
+  TTypeVariable -> SemanticTokenTypes_TypeParameter
+  -- normal data type is a tagged union type look like enum type
+  -- and a record is a product type like struct
+  -- but we don't distinguish them yet
+  TTypeCon      -> SemanticTokenTypes_Enum
+  TDataCon      -> SemanticTokenTypes_EnumMember
+  TRecField     -> SemanticTokenTypes_Property
+  -- pattern syn is like a limited version of macro of constructing a data type
+  TPatternSyn   -> SemanticTokenTypes_Macro
+  -- saturated type
+  TTypeSyn      -> SemanticTokenTypes_Type
+  -- not sure if this is correct choice
+  TTypeFamily   -> SemanticTokenTypes_Interface
+
+--   TNothing      -> Nothing
+
+lspTokenReverseMap :: Map.Map SemanticTokenTypes HsSemanticTokenType
+lspTokenReverseMap = Map.fromList $ map (\x -> (toLspTokenType x, x)) $ enumFrom minBound
+
+fromLspTokenType :: SemanticTokenTypes -> Maybe HsSemanticTokenType
+fromLspTokenType tk = Map.lookup tk lspTokenReverseMap
+
+-- * 2. Mapping from GHC type and tyThing to semantic token type.
+
+
+-- | tyThingSemantic
+tyThingSemantic :: TyThing -> Maybe HsSemanticTokenType
+tyThingSemantic ty = case ty of
+  AnId vid
+    | isTyVar vid -> Just TTypeVariable
+    | isRecordSelector vid -> Just TRecField
+    | isClassOpId vid -> Just TClassMethod
+    | isFunVar vid -> Just TFunction
+    -- \| isDFunId vid -> TClassMethod
+    | otherwise -> Just TVariable
+  AConLike con -> case con of
+    RealDataCon _ -> Just TDataCon
+    PatSynCon _   -> Just TPatternSyn
+  ATyCon tyCon
+    | isTypeSynonymTyCon tyCon -> Just TTypeSyn
+    | isTypeFamilyTyCon tyCon -> Just TTypeFamily
+    | isClassTyCon tyCon -> Just TClass
+    -- fall back to TTypeCon the result
+    | otherwise -> Just TTypeCon
+  ACoAxiom _ -> Nothing
+  where
+    isFunVar :: Var -> Bool
+    isFunVar var = isFunType $ varType var
+
+isFunType :: Type -> Bool
+isFunType a = case a of
+  ForAllTy _ t -> isFunType t
+  _x           -> isFunTy a
+
+typeSemantic :: HieKind hType -> hType -> HsSemanticTokenType
+typeSemantic kind t = case kind of
+  HieFresh -> if isFunType t then TFunction else TVariable
+  HieFromDisk full_file ->
+    if isFixFunction fullType
+      then TFunction
+      else TVariable
+    where
+      fullType = recoverFullType t (hie_types full_file)
```

Comment:
This might be an argument against trying to distinguish between things of function type and non-function type, if that's expensive :thinking: Or is this the sort of thing where instead we should consider indexing this information into `hiedb` itself?

Reply:
> Calling `recoverFullType` on every type can be very expensive, see https://gitlab.haskell.org/ghc/haddock/-/blob/b0b0e0366457c9aefebcc94df74e5de4d00e17b7/haddock-api/src/Haddock/Backends/Hyperlinker/Utils.hs#L107 for an implementation with sharing.
> 
> 
> 
> I imagine you want to map `hie_types` to something like `A.Array TypeIndex Bool`, ensuring we only look at each index once.

Sure I'll see how to arrange this in the query logic.

Reply:
cool, I'll take a look at this.

Reply:
adjusted